### ISSUE TITLE
Whisper bug!

### ISSUE THREAD

**Issue by Tortoise17** [DakeQQ/Transcribe-and-Translate-Subtitles#3]
With the default settings, it misses several sentences or words. Something is messed up with the silence / VAD and whisper. Normal use of whisper recovers much more. Is there any way to overcome this problem with the missing parts? This is something not the engine problem, but the VAD and silence settings.

**Comment by Tortoise17** [comment]
@DakeQQ if we can use the default VAD, quite likely that the quality becomes better.

**Comment by DakeQQ** [comment]
Yes, the issue you're encountering is a known problem. We have observed that sometimes Whisper 'doesn't talk' due to over-denoising or inappropriate voice activity detection (VAD) parameter settings. It's challenging to use default settings that fit all video inputs, as most defaults are taken from the official settings. We are actively working on finding a general solution. Thank you for bringing this issue to our attention.

The default VAD (timestamps generated by Whisper itself) is very time-consuming, a challenge also faced by other well-known projects. Therefore, we decided to use Official-Silero or Faster_Whisper-Silero with denoiser instead. It might be the right moment to choose between speed and accuracy.

By the way, Could you provide us with a test audio file to help reproduce the issue you're encountering?

**Comment by DakeQQ** [comment]
@Tortoise17 
You can try setting the VAD slider bars for 'Voice State Score' and 'Silence State Score' close to zero, which increases VAD sensitivity, allowing the Whisper model to 'talk more'.

**Comment by Tortoise17** [comment]
@DakeQQ I did it and tried to get maximum. Still something between silence and VAD gets overlapped.

**Comment by DakeQQ** [comment]
We have updated another powerful VAD [Pyannote-Segmentation-3.0](https://huggingface.co/pyannote/segmentation-3.0). You can follow the instructions in the README.md file to download the model & latest run.py and give it a try.

**Comment by Tortoise17** [comment]
@DakeQQ Really hats off. Great integration. !! I will let you know about the quality. Of course, the performance efficiency declined but quality measures will be shared soon. Really a great help.

**Comment by Tortoise17** [comment]
@DakeQQ If you can just add whisper large-v2 and large model / medium model option to the pipeline, this may help also in the efficiency because the segmentation of pyannote-audio is much better now. I guess. !! Still I have to make some tests until I can provide you a proper feedback. But still a big thank you for this addition. For several files, now denoising is not required. This repo is working quite fast.

**Comment by DakeQQ** [comment]
The dropdown list includes an option called 'Custom Whisper.' This allows user to use others Whisper, such as V2 or medium model using this [repository](https://github.com/DakeQQ/Automatic-Speech-Recognition-ASR-ONNX) and specify the path to the exported model. If you are interested in any specific Whisper model, let us know. We can export and upload it to the cloud, provided there is sufficient free storage space.

**Comment by Tortoise17** [comment]
@DakeQQ Thank you for a great help. Let me figure out some tests and I will share my ideas with you. May take some days. Will keep in touch.

**Comment by Tortoise17** [comment]
@DakeQQ now after several updates, now timestamps of the transcript with pyannote.audio segmentation is worse. Is there any specific parameter to set? Normally segmentation should not affect this.

**Comment by DakeQQ** [comment]
@Tortoise17 
Apologies for the confusion caused by the frequent updates. The issue may stem from the VAD padding time, which is an important parameter affecting timestamp accuracy. In the previous version of `run.py`, this was set to a constant, 800 ms. In the latest version, we have exposed this parameter for users to configure. Please download the latest version and try it yourself.

Note: In the original Pyannote-3.0 repository, the padding option was not supported. However, we encountered poor results when testing with some English movies, so we added the padding functionality. You can set it to zero to disable padding.

**Comment by Tortoise17** [comment]
@DakeQQ if we dont use VAD for whisper, the timestamps are relatively better. Whisper don't need VAD for ASR, it needs pyannote.audio for diarization. For whisper option, omit the VAD option and see the results. Still there is slight error, I will look from where it somes. But, please have a look in that and addition of no VAD as optional will make it flexible and  improve significantly the results. Also, add the word level export as json which whisper has by default with vtt export option.

**Comment by DakeQQ** [comment]
Thank you for your insights. We’d like to clarify that Pyannote-Segmentation-3.0 is a type of VAD, as it detects speech and outputs voice segments. VAD is crucial for transcription, ensuring that only relevant audio is processed efficiently.

You mentioned “Whisper doesn’t need VAD for ASR”, but Whisper uses the "no_speech" prompt to detect silence. However, applying this feature requires a sliding window approach, where the model performs inference across all audio segments to check for silence. This method is highly time-consuming and computationally expensive.

Additionally, the word-level timestamps seen in other implementations are not directly output by Whisper. These require significant post-processing of vocab logits over a sliding window.

In summary, your suggestions seems to center around applying the sliding window method. While feasible, it is expected to be time-consuming. Other well-known projects, such as Faster-Whisper, already support this. We prefer to continue using various powerful third-party VAD models like Pyannote-Segmentation-3.0, Silero... instead of relying solely on Whisper’s built-in VAD.

Best regards

**Comment by Tortoise17** [comment]
`Additionally, the word-level timestamps seen in other implementations are not directly output by Whisper. These require significant post-processing of vocab logits over a sliding window.`

@DakeQQ  Is it possible to add this possibility here? would be a great addition.

**Comment by DakeQQ** [comment]
You're probably looking for timestamps like the ones below, right?

Unfortunately, it requires a sliding window across the entire audio with Whisper's VAD no-speech detection. Otherwise, only God knows which voice corresponds to which word at what timestamp. Common VADs like Pyannote-Segmentation-3.0 and Silero don’t support this functionality yet.

```
{
  "text": " Hello! How are you?",
  "segments": [
    {
      "id": 0,
      "seek": 0,
      "start": 0.5,
      "end": 1.2,
      "text": " Hello!",
      "tokens": [ 25431, 2298 ],
      "temperature": 0.0,
      "avg_logprob": -0.6674491882324218,
      "compression_ratio": 0.8181818181818182,
      "no_speech_prob": 0.10241222381591797,
      "confidence": 0.51,
      "words": [
        {
          "text": "Hello!",
          "start": 0.5,
          "end": 1.2,
          "confidence": 0.51
        }
      ]
    },
    {
      "id": 1,
      "seek": 200,
      "start": 2.02,
      "end": 4.48,
      "text": " How are you?",
      "tokens": [ 50364, 4410, 12 ],
      "temperature": 0.0,
      "avg_logprob": -0.43492694334550336,
      "compression_ratio": 0.7714285714285715,
      "no_speech_prob": 0.06502953916788101,
      "confidence": 0.595,
      "words": [
        {
          "text": "How",
          "start": 2.02,
          "end": 3.78,
          "confidence": 0.441
        },
        {
          "text": "are",
          "start": 3.78,
          "end": 3.84,
          "confidence": 0.948
        },
        {
          "text": "you",
          "start": 3.84,
          "end": 4.0,
          "confidence": 0.935
        }
      ]
    }
  ],
  "language": "en"
}
```
